[cmake] allow to override the cxx standard

fabiencastan opened the pull request:
When using multiple libraries using eigen, they all need to agree on the way we use memory alignment in eigen.
See https://eigen.tuxfamily.org/dox/group__TopicStlContainers.html
As we have moved to C++17, we need all libraries using eigen to be compatible with that.

The solution is simple: define a cmake cached variable, so it can be overriden.

This PR also add some missing includes, needed to build with g++ 10.

fabiencastan committed:
[cmake] option to override cxx standard

from cmake command line

fabiencastan committed:
build fix: add missing cassert include

simogasp commented on CMakeLists.txt:
```diff
@@ -181,8 +184,6 @@ add_library( random_generators test/random_generators.cpp test/random_generators
 set_target_properties( opengv random_generators PROPERTIES
                     SOVERSION ${PROJECT_VERSION}
                     VERSION ${PROJECT_VERSION}
-                    CXX_STANDARD 11
```
same thing at line 333 below for the tests

simogasp commented on CMakeLists.txt:
```diff
@@ -181,8 +184,6 @@ add_library( random_generators test/random_generators.cpp test/random_generators
                     VERSION ${PROJECT_VERSION}
-                    CXX_STANDARD 11
-                    CXX_STANDARD_REQUIRED ON
                     DEBUG_POSTFIX d )
```
```suggestion
                     )
```
For `DEBUG_POSTFIX`we can just set a cache variable that initialize `CMAKE_DEBUG_POSTFIX` that will automatically set it for all targets, without doing it explicitly:
```cmake
set(CMAKE_DEBUG_POSTFIX "d" CACHE STRING "The postfix to use for the debug version of the target names")
```

fabiencastan committed:
[cmake] gcc flags: avoid usage of "-march"

from gcc documentation: "-march=cpu-type allows GCC to generate code
that may not run at all on processors other than the one indicated."
